tests/test_connector.rs

```diff
@@ -0,0 +1,351 @@
+    );
+    res.assert_success();
+}
+
+#[allow(dead_code)]
+fn print_logs(logs: &Vec<String>) {
+    for l in logs {
+        println!("[log] {}", l);
+    }
+}
+
+fn call_deposit_eth(master_account: &UserAccount) {
+    let proof: Proof = serde_json::from_str(PROOF_DATA_ETH).unwrap();
+    let data = DepositEthCallArgs {
+        relayer_eth_account: validate_eth_address("09f7219e434EAA7021Ae5f9Ecd0CaBc2405447A3"),
+        proof,
+    }
+    .try_to_vec()
+    .unwrap();
+    let res = master_account.call(
+        CONTRACT_ACC.to_string(),
+        "deposit_eth",
+        &data[..],
+        DEFAULT_GAS,
+        0,
+    );
+    res.assert_success();
+    //println!("{:#?}", res.promise_results());
+    //print_logs(res.logs());
+}
+
+fn get_near_balance(master_account: &UserAccount, acc: &str) -> u128 {
+    let balance = master_account.view(
+        CONTRACT_ACC.to_string(),
+        "ft_balance_of",
+        json!({ "account_id": acc }).to_string().as_bytes(),
+    );
+    String::from_utf8(balance.unwrap())
+        .unwrap()
+        .parse()
+        .unwrap()
+}
+
+fn get_eth_balance(master_account: &UserAccount, address: EthAddress) -> u128 {
+    let address = hex::encode(address);
+    let balance = master_account.view(
+        CONTRACT_ACC.to_string(),
+        "ft_balance_of_eth",
+        json!({ "address": address }).to_string().as_bytes(),
+    );
+    String::from_utf8(balance.unwrap())
+        .unwrap()
+        .parse()
+        .unwrap()
+}
+
+fn total_supply(master_account: &UserAccount) -> u128 {
+    let balance = master_account.view(CONTRACT_ACC.to_string(), "ft_total_supply", &[]);
+    String::from_utf8(balance.unwrap())
+        .unwrap()
+        .parse()
+        .unwrap()
+}
+
+fn total_supply_near(master_account: &UserAccount) -> u128 {
+    let balance = master_account.view(CONTRACT_ACC.to_string(), "ft_total_supply_near", &[]);
+    String::from_utf8(balance.unwrap())
+        .unwrap()
+        .parse()
+        .unwrap()
+}
+
+fn total_supply_eth(master_account: &UserAccount) -> u128 {
+    let balance = master_account.view(CONTRACT_ACC.to_string(), "ft_total_supply_eth", &[]);
+    String::from_utf8(balance.unwrap())
+        .unwrap()
+        .parse()
+        .unwrap()
+}
+
+#[test]
+fn test_near_deposit_balance_total_supply() {
+    let (master_account, _contract) = init(CUSTODIAN_ADDRESS);
+    call_deposit_near(&master_account);
+
+    let balance = get_near_balance(&master_account, DEPOSITED_RECIPIENT);
+    assert_eq!(balance, DEPOSITED_AMOUNT - DEPOSITED_FEE);
+
+    let balance = get_near_balance(&master_account, CONTRACT_ACC);
+    assert_eq!(balance, DEPOSITED_FEE);
+
+    let balance = total_supply(&master_account);
+    assert_eq!(balance, DEPOSITED_AMOUNT);
+
+    let balance = total_supply_near(&master_account);
+    assert_eq!(balance, DEPOSITED_AMOUNT);
+
+    let balance = total_supply_eth(&master_account);
+    assert_eq!(balance, 0);
+}
+
+#[test]
+fn test_deposit_eth_and_near() {
```

Comment:
I am uncertain about this test. It calls `init` twice, so I am unsure what happens to the state in between. I'm also not clear on why we need two different custodian addresses between these calls.

I would expect this test to be similar to the ones above and below; it initializes the contract, deposits some near and eth, then checks the all the total supplies are the right numbers. Is there a complication here I am missing?

Reply:
We have different Proof and EIP712 messages that related to different custodian_address. Kirill can't generate data for the same custodian address. That's why we call `init` again.